ambry-protocol/src/test/java/com.github.ambry.protocol/RequestResponseTest.java

```diff
@@ -133,17 +136,29 @@ public long writeTo(WritableByteChannel channel)
 }
 
 public class RequestResponseTest {
+  private Random random = new Random();
+
   private void testPutRequest(MockClusterMap clusterMap, int correlationId, String clientId, BlobId blobId,
       BlobProperties blobProperties, byte[] userMetadata, BlobType blobType, byte[] blob, int blobSize)
       throws IOException {
     int sizeInBlobProperties = (int) blobProperties.getBlobSize();
     PutRequest request = new PutRequest(correlationId, clientId, blobId, blobProperties, ByteBuffer.wrap(userMetadata),
         new ByteBufferInputStream(ByteBuffer.wrap(blob)), blobSize, blobType);
     ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-    WritableByteChannel writableByteChannel = Channels.newChannel(outputStream);
+    // Create channels in such a way that writeTo() may not be able to write out all the data at once.
+    int allocationSize = random.nextInt((int) request.sizeInBytes()) + 1;
```

Comment:
Random only might not be good for the test. The test should have a case where `writeTo()` *surely* cannot write out all data at once. In the current setup, there is a chance that `allocationSize == request.sizeInBytes()`. This means that buggy code *might* pass the test. I would prefer to have the test explicitly tests the cases <=> and then random.